extensions/ql-vscode/src/test-adapter.ts

```diff
@@ -182,11 +184,33 @@ export class QLTestAdapter extends DisposableObject implements TestAdapter {
     testLogger.outputChannel.show(true);
 
     this.runningTask = this.track(new CancellationTokenSource());
+    const token = this.runningTask.token;
 
     this._testStates.fire({ type: 'started', tests: tests } as TestRunStartedEvent);
 
+    const currentDatabaseUri = this.databaseManager.currentDatabaseItem?.databaseUri;
     try {
-      await this.runTests(tests, this.runningTask.token);
+      const databasesUnderTest = this.databaseManager.databaseItems
+        .filter(database => tests.find(testPath => database.isAffectedByTestsUnder(testPath)));
+      for (const database of databasesUnderTest) {
+        await this.databaseManager
+          .removeDatabaseItem(_ => { /* no progress reporting */ }, token, database);
+      }
+      await this.runTests(tests, token);
+      for (const closedDatabase of databasesUnderTest) {
+        try {
+          const uri = closedDatabase.databaseUri;
+          const reopenedDatabase = await this.databaseManager
+            .openDatabase(_ => { /* no progress reporting */ }, token, uri);
+          await this.databaseManager.renameDatabaseItem(reopenedDatabase, closedDatabase.name);
+          if (currentDatabaseUri === reopenedDatabase.databaseUri) {
+            await this.databaseManager.setCurrentDatabaseItem(reopenedDatabase, true);
+          }
+        } catch (e) {
+          // The database directory may have been deleted by a successful test run.  So failures
```

Comment:
Is there a specific error message we can check here for missing databases? Seems like we only want to catch errors relating to missing databases and rethrow all others.

Reply:
What do you think we should do if multiple databases fail to open for non-missing errors?  That would lead to multiple exceptions that need to be propagated.  Is there a way to throw multiple exceptions simultaneously in Typescript?  Do we stop at the first exception and throw that?  Or do something else?

Reply:
That's a good question. There are two possibilities that I can think of:

1. keep track of all catastrophic errors, concatenate the error messages and throw after trying to re-open all old databases
2. fail fast after the first error, throw that one and don't try to add the remainder.

While 1) may be nicer, I expect that 2) may be sufficient because hopefully this will be a rare occurrence, users are unlikely to have more than one testproj database they are working on at once, and if there is a catastrophic error it's likely that it would affect all of the databases at the same time.